lib/ipam/ipam.go

```diff
@@ -1383,6 +1344,62 @@ func (c ipamClient) convertBackendToIPAMConfig(cfg *model.IPAMConfig) *IPAMConfi
 	}
 }
 
+// maybeReleaseAffinity retrieves the pool and node for the given block and determines
+// if the pool still selects node. If it no longer matches, it will release the block
+// affinity for that node.
+// Returns a bool indicating if the block was released.
+func (c ipamClient) maybeReleaseAffinity(ctx context.Context, b *model.AllocationBlock) (bool, error) {
+	logCtx := log.WithField("cidr", b.CIDR)
+	// Retrieve node for this allocation. We do this so we can clean up affinity for blocks
+	// which should no longer be affine to this host.
+	host := getHostAffinity(b)
+	logCtx.Debugf("block affinity: %s", host)
+
+	// Exit function is host is empty.
+	if host == "" {
+		return false, nil
+	}
+
+	// If the IP pool which owns this block no longer selects this node,
+	// we should release the block's affinity to this node so it can be
+	// used elsewhere.
+	logCtx.Debugf("Looking up node for host affinity %s", host)
```

Comment:
I'd clarify we're looking up the node _labels_ here - it will help whoever is reading these logs to keep the context of what's going on.